[core] (cgroups 2/n) adding integration tests for the cgroup sysfs driver.

israbbani opened the pull request:
This PR adds integration tests for the `SysFsCgroupDriver`. 
This PR stacks on [#54898]. For more details about the resource isolation project see #54703. 

The design goals of the tests are:
1. they are fast and repeatable (i.e. not flaky)
2. they don't leak any state (cgroups, directories, processes)
3. they can be run inside a VM (for local development) or in a container (with escalated privileges in CI)

The tests make the following assumptions:
1. Cgroupv2 is mounted in [unified mode](https://kubernetes.io/docs/concepts/architecture/cgroups/#linux-distribution-cgroup-v2-support) i.e. cgroup1 is disabled.
2. The user that starts the entrypoint script `sysfs_cgroup_driver_integration_test_entrypoint.sh` has read-write access to the ROOT_CGROUP (which is defined by a variable inside `sysfs_cgroup_driver_integration_test_entrypoint.sh`)
3. The user has permissions to create a new user (called cgroup-tester) to be able to run tests with different privilege levels. This is useful e.g. when you run the setup/teardown as the root user, but want to code paths that check for permissions and require an unprivileged user to test the unhappy path (i.e. user does not have permissions to a cgroup and a write fails)

The tests created scoped cgroup hierarchy to simplify cleanup and isolate test runs from each other. This looks like:
```bash
#                        ROOT_CGROUP
#                             |
#                        BASE_CGROUP
#                       /           \
#                 TEST_CGROUP   LEAF_CGROUP
```
Where
* ROOT_CGROUP is provided by the environment running the test (for CI it's /sys/fs/cgroup)
* BASE_CGROUP is a random cgroup created for the test suite (e.g. /sys/fs/cgroup/testing.axUYw
* TEST_CGROUP is the cgroup used by the integration tests
* LEAF_CGROUP is used to migrate all processes in the ROOT_CGROUP (if it has any) so that it can be allowed to have child cgroups and still pass the no internal processes constraint.

The tests follow a peculiar CI pattern:
* The endpoint is triggered directly from the buildkite CI script (instead of through bazel). This is because if you run the endpoint as a bazel `sh_test` then, no other user has access to the bazel workspace and we cannot run the setup/teardown as a privileged user and the cpp tests as an unprivileged user. 

Example output from a test run:
```
Starting Cgroupv2 Integration Tests as user root
ROOT_CGROUP is /sys/fs/cgroup.
ROOT_CGROUP is /sys/fs/cgroup.
BASE_CGROUP for the test suite is /sys/fs/cgroup/testing.ydOxJ.
TEST_CGROUP for the test suite is /sys/fs/cgroup/testing.ydOxJ/test.
LEAF_CGROUP for the test suite is /sys/fs/cgroup/testing.ydOxJ/leaf.
Starting integration test fixture with:
  ACTION=setup
  ROOT_CGROUP=/sys/fs/cgroup
  BASE_CGROUP=/sys/fs/cgroup/testing.ydOxJ
  TEST_CGROUP=/sys/fs/cgroup/testing.ydOxJ/test
  UNPRIV_USER=cgroup-tester
Running ACTION: setup
Created LEAF_CGROUP at /sys/fs/cgroup/testing.ydOxJ/leaf.
Created TEST_CGROUP at /sys/fs/cgroup/testing.ydOxJ/test.
Moved 3 procs from /sys/fs/cgroup/cgroup.procs to /sys/fs/cgroup/testing.ydOxJ/leaf/cgroup.procs.
Updated +cpu +memory controllers for /sys/fs/cgroup/cgroup.subtree_control
Updated +cpu +memory controllers for /sys/fs/cgroup/testing.ydOxJ/cgroup.subtree_control
Updated +cpu +memory controllers for /sys/fs/cgroup/testing.ydOxJ/test/cgroup.subtree_control
Created unprivilged user cgroup-tester.
cgroup-tester is the owner the cgroup subtree starting at /sys/fs/cgroup/testing.ydOxJ
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //src/ray/common/cgroup2/tests:sysfs_cgroup_driver_integration_test
-----------------------------------------------------------------------------
Running main() from gmock_main.cc
[==========] Running 36 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 36 tests from SysFsCgroupDriverIntegrationTest
[ RUN      ] SysFsCgroupDriverIntegrationTest.SysFsCgroupDriverIntegrationTestFailsIfNoCgroupTestPathSpecified
[       OK ] SysFsCgroupDriverIntegrationTest.SysFsCgroupDriverIntegrationTestFailsIfNoCgroupTestPathSpecified (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.CheckCgroupFailsIfCgroupv2PathButNoReadPermissions
[       OK ] SysFsCgroupDriverIntegrationTest.CheckCgroupFailsIfCgroupv2PathButNoReadPermissions (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.CheckCgroupFailsIfCgroupv2PathButNoWritePermissions
[       OK ] SysFsCgroupDriverIntegrationTest.CheckCgroupFailsIfCgroupv2PathButNoWritePermissions (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.CheckCgroupFailsIfCgroupv2PathButNoExecPermissions
[       OK ] SysFsCgroupDriverIntegrationTest.CheckCgroupFailsIfCgroupv2PathButNoExecPermissions (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.CheckCgroupSucceedsIfCgroupv2PathAndReadWriteExecPermissions
[       OK ] SysFsCgroupDriverIntegrationTest.CheckCgroupSucceedsIfCgroupv2PathAndReadWriteExecPermissions (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.CreateCgroupFailsIfAlreadyExists
[       OK ] SysFsCgroupDriverIntegrationTest.CreateCgroupFailsIfAlreadyExists (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.CreateCgroupFailsIfAncestorCgroupDoesNotExist
[       OK ] SysFsCgroupDriverIntegrationTest.CreateCgroupFailsIfAncestorCgroupDoesNotExist (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.CreateCgroupFailsIfOnlyReadPermissions
[       OK ] SysFsCgroupDriverIntegrationTest.CreateCgroupFailsIfOnlyReadPermissions (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.CreateCgroupFailsIfOnlyReadWritePermissions
[       OK ] SysFsCgroupDriverIntegrationTest.CreateCgroupFailsIfOnlyReadWritePermissions (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.CreateCgroupSucceedsIfParentExistsAndReadWriteExecPermissions
[       OK ] SysFsCgroupDriverIntegrationTest.CreateCgroupSucceedsIfParentExistsAndReadWriteExecPermissions (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.GetAvailableControllersFailsIfCgroupDoesNotExist
[       OK ] SysFsCgroupDriverIntegrationTest.GetAvailableControllersFailsIfCgroupDoesNotExist (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.GetAvailableControllersFailsIfReadWriteButNotExecutePermissions
[       OK ] SysFsCgroupDriverIntegrationTest.GetAvailableControllersFailsIfReadWriteButNotExecutePermissions (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.GetAvailableControllersSucceedsWithCPUAndMemoryControllersOnBaseCgroup
[       OK ] SysFsCgroupDriverIntegrationTest.GetAvailableControllersSucceedsWithCPUAndMemoryControllersOnBaseCgroup (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.GetAvailableControllersSucceedsWithNoAvailableControllers
[       OK ] SysFsCgroupDriverIntegrationTest.GetAvailableControllersSucceedsWithNoAvailableControllers (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.MoveAllProcessesFailsIfSourceDoesntExist
[       OK ] SysFsCgroupDriverIntegrationTest.MoveAllProcessesFailsIfSourceDoesntExist (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.MoveAllProcessesFailsIfDestDoesntExist
[       OK ] SysFsCgroupDriverIntegrationTest.MoveAllProcessesFailsIfDestDoesntExist (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.MoveAllProcessesFailsIfNotReadWriteExecPermissionsForSource
[       OK ] SysFsCgroupDriverIntegrationTest.MoveAllProcessesFailsIfNotReadWriteExecPermissionsForSource (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.MoveAllProcessesFailsIfNotReadWriteExecPermissionsForDest
[       OK ] SysFsCgroupDriverIntegrationTest.MoveAllProcessesFailsIfNotReadWriteExecPermissionsForDest (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.MoveAllProcessesFailsIfNotReadWriteExecPermissionsForAncestor
[       OK ] SysFsCgroupDriverIntegrationTest.MoveAllProcessesFailsIfNotReadWriteExecPermissionsForAncestor (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.MoveAllProcessesSucceedsWithCorrectPermissionsAndValidCgroups
[       OK ] SysFsCgroupDriverIntegrationTest.MoveAllProcessesSucceedsWithCorrectPermissionsAndValidCgroups (15 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.EnableControllerFailsIfReadOnlyPermissionsForCgroup
[       OK ] SysFsCgroupDriverIntegrationTest.EnableControllerFailsIfReadOnlyPermissionsForCgroup (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.EnableControllerFailsIfReadWriteOnlyPermissionsForCgroup
[       OK ] SysFsCgroupDriverIntegrationTest.EnableControllerFailsIfReadWriteOnlyPermissionsForCgroup (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.EnableControllerFailsIfCgroupDoesNotExist
[       OK ] SysFsCgroupDriverIntegrationTest.EnableControllerFailsIfCgroupDoesNotExist (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.EnableControllerFailsIfControllerNotAvailableForCgroup
[       OK ] SysFsCgroupDriverIntegrationTest.EnableControllerFailsIfControllerNotAvailableForCgroup (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.DisableControllerFailsIfControllerNotEnabled
[       OK ] SysFsCgroupDriverIntegrationTest.DisableControllerFailsIfControllerNotEnabled (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.DisableControllerFailsIfReadOnlyPermissionsForCgroup
[       OK ] SysFsCgroupDriverIntegrationTest.DisableControllerFailsIfReadOnlyPermissionsForCgroup (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.DisableControllerFailsIfReadWriteOnlyPermissionsForCgroup
[       OK ] SysFsCgroupDriverIntegrationTest.DisableControllerFailsIfReadWriteOnlyPermissionsForCgroup (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.DisableControllerFailsIfCgroupDoesNotExist
[       OK ] SysFsCgroupDriverIntegrationTest.DisableControllerFailsIfCgroupDoesNotExist (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.EnableAndDisableControllerSucceedWithCorrectInputAndPermissions
[       OK ] SysFsCgroupDriverIntegrationTest.EnableAndDisableControllerSucceedWithCorrectInputAndPermissions (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.AddResourceConstraintFailsIfCgroupDoesntExist
[       OK ] SysFsCgroupDriverIntegrationTest.AddResourceConstraintFailsIfCgroupDoesntExist (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.AddResourceConstraintFailsIfReadOnlyPermissions
[       OK ] SysFsCgroupDriverIntegrationTest.AddResourceConstraintFailsIfReadOnlyPermissions (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.AddResourceConstraintFailsIfReadWriteOnlyPermissions
[       OK ] SysFsCgroupDriverIntegrationTest.AddResourceConstraintFailsIfReadWriteOnlyPermissions (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.AddResourceConstraintFailsIfConstraintNotSupported
[       OK ] SysFsCgroupDriverIntegrationTest.AddResourceConstraintFailsIfConstraintNotSupported (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.AddResourceConstraintFailsIfControllerNotEnabled
[       OK ] SysFsCgroupDriverIntegrationTest.AddResourceConstraintFailsIfControllerNotEnabled (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.AddResourceConstraintFailsIfInvalidConstraintValue
[       OK ] SysFsCgroupDriverIntegrationTest.AddResourceConstraintFailsIfInvalidConstraintValue (0 ms)
[ RUN      ] SysFsCgroupDriverIntegrationTest.AddResourceConstraintSucceeds
[       OK ] SysFsCgroupDriverIntegrationTest.AddResourceConstraintSucceeds (0 ms)
[----------] 36 tests from SysFsCgroupDriverIntegrationTest (21 ms total)

[----------] Global test environment tear-down
[==========] 36 tests from 1 test suite ran. (21 ms total)
[  PASSED  ] 36 tests.
Starting integration test fixture with:
  ACTION=teardown
  ROOT_CGROUP=/sys/fs/cgroup
  BASE_CGROUP=/sys/fs/cgroup/testing.ydOxJ
  TEST_CGROUP=/sys/fs/cgroup/testing.ydOxJ/test
  UNPRIV_USER=cgroup-tester
Running ACTION: teardown
Looking for files to backup/remove ...
Removing files ...
Removing user `cgroup-tester' ...
Warning: group `cgroup-tester' has no more members.
Done.
Deleted unprivilged user cgroup-tester.
Updated -cpu -memory controllers for /sys/fs/cgroup/testing.ydOxJ/test/cgroup.subtree_control
Updated -cpu -memory controllers for /sys/fs/cgroup/testing.ydOxJ/cgroup.subtree_control
Updated -cpu -memory controllers for /sys/fs/cgroup/cgroup.subtree_control
Moved 4 procs from /sys/fs/cgroup/testing.ydOxJ/leaf/cgroup.procs to /sys/fs/cgroup/cgroup.procs.
Deleted /sys/fs/cgroup/testing.ydOxJ/test
Deleted /sys/fs/cgroup/testing.ydOxJ/leaf
Deleted /sys/fs/cgroup/testing.ydOxJ
Teardown successful.
```

Commit:
[core] (cgroups 1/n) Adding a sys/fs filesystem driver
to perform cgroup operations.

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
adding the copyright

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Adding a fallback for creating processes inside cgroups with fork/exec
instead of clone for older kernel headers < 5.7 (which is what we have
in CI)

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
adding a pause in the tests to see what's up with the container

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Update src/ray/common/cgroup2/cgroup_driver_interface.h

Co-authored-by: Edward Oakes <ed.nmi.oakes@gmail.com>
Signed-off-by: Ibrahim Rabbani <israbbani@gmail.com>

Commit:
Comments

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Merge branch 'irabbani/cgroups-1' of github.com:ray-project/ray into irabbani/cgroups-1

Commit:
Putting the cgroupv2 tests into a separate target

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
removing test sleep

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Removing a docstring

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
enabling CI tests

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
fixing absl imports

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
commenting local

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
doxygen formatting

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Merge branch 'master' into irabbani/cgroups-1

Commit:
removing integration tests

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
final cleanup

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
iwyu

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Merge branch 'master' into irabbani/cgroups-1

Commit:
we cpplintin!

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Update src/ray/common/cgroup2/sysfs_cgroup_driver.cc

Co-authored-by: Edward Oakes <ed.nmi.oakes@gmail.com>
Signed-off-by: Ibrahim Rabbani <israbbani@gmail.com>

Commit:
Apply suggestions from code review

Co-authored-by: Edward Oakes <ed.nmi.oakes@gmail.com>
Signed-off-by: Ibrahim Rabbani <israbbani@gmail.com>

Commit:
bug

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Merge branch 'irabbani/cgroups-1' of github.com:ray-project/ray into irabbani/cgroups-1

Commit:
[core] Integration tests for SysFsCgroupDriver.

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Cleaning up cgroup_test_utils and attempting to
fix CI.

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
broken

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
up

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
upup

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Merge branch 'master' into irabbani/cgroups-2

Signed-off-by: Ibrahim Rabbani <irabbani@anyscale.com>

Commit:
up

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Adding shell scripts to do cgroup setup/teardown
and enabling test for ci

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Merge branch 'master' into irabbani/cgroups-2

Commit:
Merged and fixed a few issues

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
fixing test target for CI

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
maybe this will trigger tests

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
runforever

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
up

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
up

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
cleaning up todos and docs

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
one more

Signed-off-by: irabbani <irabbani@anyscale.com>

israbbani commented on src/ray/common/cgroup2/sysfs_cgroup_driver.cc:
```diff
@@ -56,8 +56,8 @@ Status SysFsCgroupDriver::CheckCgroupv2Enabled() {
   while ((mnt = getmntent(fp)) != nullptr) {
-    found_cgroupv1 = found_cgroupv1 || strcmp(mnt->mnt_fsname, "cgroup") == 0;
-    found_cgroupv2 = found_cgroupv2 || strcmp(mnt->mnt_fsname, "cgroup2") == 0;
+    found_cgroupv1 = found_cgroupv1 || strcmp(mnt->mnt_type, "cgroup") == 0;
+    found_cgroupv2 = found_cgroupv2 || strcmp(mnt->mnt_type, "cgroup2") == 0;
```
This was a bug. It was caught by the integration tests. I fixed the unit test to catch it too now.

Commit:
adding separate target for unit tests now

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
typo

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
come unit and integration test targets

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
missing flag

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
plz work

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
one more

Signed-off-by: irabbani <irabbani@anyscale.com>

israbbani commented on .buildkite/core.rayci.yml:
```diff
@@ -303,15 +303,15 @@ steps:
-      - bazel run //ci/ray_ci:test_in_docker -- //:all //src/... core --only-tags=cgroup --build-type cgroup
-        --privileged --cache-test-results
-
+      - bazel run //ci/ray_ci:test_in_docker -- //:all //src/ray/common/cgroup2/tests:sysfs_cgroup_driver_test
```
I'm currently putting in the full test name. I'll clean up the test targets once I delete the old cgroup code. I added a TODO in https://github.com/ray-project/ray/issues/54703

edoakes commented on src/ray/common/cgroup2/tests/BUILD.bazel:
```diff
@@ -31,3 +32,22 @@ ray_cc_test(
     ],
 )
+
+# This test is run through sysfs_cgroup_driver_integration_test_entrypoint.sh
```
leave instructions for how to run them manually (if they're in the shell script, just mention it here)

edoakes commented on src/ray/common/cgroup2/tests/BUILD.bazel:
```diff
@@ -31,3 +32,22 @@ ray_cc_test(
+    srcs = ["sysfs_cgroup_driver_integration_test.cc"],
+    tags = [
+        "cgroup",
+        "no_windows",
```
I think you can use the bazel flag I sent you on slack to have this only run on linux instead of blacklisting windows

edoakes commented on src/ray/common/cgroup2/tests/cgroup_test_utils.cc:
```diff
@@ -35,44 +35,239 @@
 #include "ray/common/status_or.h"
 #include "ray/util/logging.h"
 
+std::string GenerateRandomFilename(size_t len) {
```
just use `UniqueID::FromRandom()`, that's what we do elsewhere

edoakes replied:
if `id.h` is slow that is quite bad :0

might be some easy wins there...

edoakes reviewed (commented):
Looks good. Please write instructions somewhere discoverable for how someone should run these locally.

israbbani commented on src/ray/common/cgroup2/tests/BUILD.bazel:
```diff
@@ -31,3 +32,22 @@ ray_cc_test(
 )
+
+# This test is run through sysfs_cgroup_driver_integration_test_entrypoint.sh
+ray_cc_test(
```
Makes sense. What about the `cgroup_test_utils.{h|cc}` files?

Commit:
Addressing feedback
- integration tests get their own directory
- using bazel supported platforms instead of a no_windows tag (which we
  can probably deprecate with this flag)

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
ci change

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Begrudingly using the random id generator from id.h

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
instructions for running locally

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
adding instructions to run locally

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Merge branch 'master' into irabbani/cgroups-2

Commit:
fixing ci

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Merge branch 'irabbani/cgroups-2' of github.com:ray-project/ray into irabbani/cgroups-2

Commit:
ci

Signed-off-by: irabbani <irabbani@anyscale.com>
